urbit/app/eth-event-viewer.hoon

```diff
@@ -0,0 +1,409 @@
+  :~  (leave-eth-watcher address.act)
+      (watch-eth-watcher address.act)
+  ==
+::
+++  handle-get-abi
+  |=  act=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?>  ?=(%get-abi -.act)
+  :_  state
+  [(request-ethereum-abi address.act) ~]
+::
+++  handle-add-contract
+  |=  act=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?>  ?=(%add-contract -.act)
+  ?:  =(%.y (~(has by contracts.state) address.contract.act))
+      [~ state]
+  =/  address  address.contract.act
+  :_  state(contracts (~(put by contracts.state) address contract.act))
+  :~  (request-ethereum-abi address.contract.act)
+      [%give %fact [/state/update ~] %json !>((new-json contract.act))]
+      (setup-eth-watcher contract.act)
+      (watch-eth-watcher address.contract.act)
+  ==
+::
+++  new-json
+  |=  new-contract=contract-type
+  ^-  json
+  =,  enjs:format
+  %-  pairs
+  [%new-contract (contract-encoder new-contract)]~
+  
+::
+++  handle-remove-contract
+  |=  act=eth-event-viewer-action
+  ^-  (quip card _state)
+  ?>  ?=(%remove-contract -.act)
+  :_  state(contracts (~(del by contracts.state) address.act))
+  :~  [%give %fact [/state/update ~] %json !>((remove-json address.act))]
+      (leave-eth-watcher address.act)
+      (clear-eth-watcher address.act)
+  ==
+::
+++  remove-json
+  |=  address=@ux
+  ^-  json
+  =,  enjs:format
+  %-  pairs
+  :~  [%remove-contract (tape (trip (ux-to-cord address)))]
+  ==
+::
+++  contracts-json
+  |=  =_state
+  ^-  json
+  =,  enjs:format
+  %-  pairs
+  :~  [%contracts (contracts-encoder state)]
+  ==
+::
+++  contracts-encoder
+  |=  =_state
+  ^-  json
+  =,  enjs:format
+  =/  contract-list
+    `(list contract-type)`~(val by contracts.state)
+  `json`a+(turn contract-list contract-list-encoder)
+::
+++  contract-list-encoder  |=(=contract-type (contract-encoder contract-type))
+::
+++  contract-encoder
+  |=  =contract-type
+  ^-  json
+  =,  enjs:format
+  %-  pairs
+  :~  [%address (tape (trip (ux-to-cord address.contract-type)))]
+      [%name (tape (trip name.contract-type))]
+      [%abi-events (tape (trip abi-events.contract-type))]
+      [%specific-events `json`a+(turn `wain`specific-events.contract-type |=(=cord s+cord))]
+      [%event-logs (event-logs-encoder:ew-lib event-logs.contract-type)]
+  ==
+::
+++  poke-handle-http-request
+  |=  =inbound-request:eyre
+  ^-  simple-payload:http
+  =+  url=(parse-request-line url.request.inbound-request)
+  ?+  site.url  not-found:gen
+      [%'~eth-event-viewer' %css %index ~]  (css-response:gen style)
+      [%'~eth-event-viewer' %js %tile ~]    (js-response:gen tile-js)
+      [%'~eth-event-viewer' %js %index ~]   (js-response:gen script)
+  ::
+      [%'~eth-event-viewer' %img @t *]
+    =/  name=@t  i.t.t.site.url
+    =/  img  (~(get by eth-event-viewer-png) name)
+    ?~  img
+      not-found:gen
+    (png-response:gen (as-octs:mimes:html u.img))
+  ::
+      [%'~eth-event-viewer' *]  (html-response:gen index)
+  ==
+::
+++  http-response
```

Comment:
So this gets the result from the requests, and then always just forwards it to the frontend.

Something about this is righteous, in that you're using your urbit as your gateway to all data you care about, and it just gets that for you however is most appropriate.

But it's also some external thing, and urbit generally tries to care about external things/dependencies as little as possible. The app logic would be a bit simpler if this request simply happened on/by the frontend, making this app's API _exclusively_ about contracts and their events, instead of also dealing with ABIs in some places.

Is there something I'm missing, or could this be safely moved to the frontend? How would you feel about that?